content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+### Basic Block Chaining
+As we reorder blocks, we will need to insert additional jumps to preserve the control flow. Therefore, we wish to order basic blocks in a way that minimizes these unconditional jumps. This is to prevent unnecessary instruction pointer moves during the execution of the program. We define a basic block chain as a directed sequence of basic blocks where no unconditional jumps occur except possibly in the last block. Generating a maximum-cardinality set of these chains would constitute the fewest number of unconditional jumps required. The authors denote this problem the *fall-through maximization problem*. This problem is the same as the maximum path cover problem, which is **NP-hard**. PH uses a greedy heuristic to approach this but does not provide a theoretical guarantee of how close to optimal the generated solution is.
+
+There *is* is a 1/2-approximation algorithm for this problem guarantees that the weight of the path cover is within *1/2* of the optimal solution. While its performance does not absolutely exceed that of PH's heuristic, Codestitcher uses a hybrid of PH's method and the approximation algorithm, which provides better results than either algorithm run individually.
+
+### Hierarchical Code Collocation
+
+Coming up with a framework to integrate intraprocedural and interprocedural ordering optimizations while taking *layout distance* (the distance between two instructions) into account is difficult. We will first try to formalize this problem.
+
+#### *d*-Close Partial Layout
+
+It would help if we could formally define what layout distance really means. The paper's definition is quite intuitive: the number of bytes between two basic blocks. To take advantage of spatial locality, we want to ensure that related instructions that are executed frequently have as few bytes between them as possible. (This is because the instruction cache can only hold so many instructions.)
+
+Given a specific distance value *d*, we can try to lay out our program to maximize the number of times control is transferred between two basic blocks within *d* bytes of each other. However, this is not enough. We need to optimize over many different *d* values to account for hierarchical memory systems with multiple cache levels. Our goal is to find the *finest grain* layout that maximizes control transfers across all combinations of basic blocks, across multiple distances.
+
+How do we solve this? The paper suggests a greedy approach that encodes the problem as a graph of basic block sequences, with edges weighted by the number of control flow transfers within *d* bytes. Nodes of the graph are iteratively coalesced in order of descending weight until no edges remain, and the remaining node's ordering is output.
```

Comment:
> and the remaining node's ordering is output.

I don't quite understand this part of the sentence—what is the "remaining node"? Maybe you mean plural "nodes"? If so, what is their ordering? If no edges are left, there isn't an ordering anymore, right?

Reply:
When two nodes are coalesced, the new node stores an internal ordering of the two original nodes. For example, if node A and node B were coalesced, the resulting node would store either A-B or B-A. Since we coalesce until no edges remain, we'll end up with one node, which will contain the final function ordering. I've updated this portion to clarify.